components/script/layout_wrapper.rs

```diff
@@ -757,46 +779,48 @@ impl<'ln> ServoThreadSafeLayoutNode<'ln> {
     }
 }
 
+// NB: The implementation here is a bit tricky because elements implementing
+// pseudos are supposed to return false for is_element().
 impl<'ln> NodeInfo for ServoThreadSafeLayoutNode<'ln> {
     fn is_element(&self) -> bool {
         self.pseudo == PseudoElementType::Normal && self.node.is_element()
     }
 
     fn is_text_node(&self) -> bool {
-        // It's unlikely that text nodes will ever be used to implement a
-        // pseudo-element, but the type system doesn't really enforce that,
-        // so we check to be safe.
-        self.pseudo == PseudoElementType::Normal && self.node.is_text_node()
+        self.node.is_text_node()
     }
 
     fn needs_layout(&self) -> bool {
-        self.pseudo != PseudoElementType::Normal ||
-        self.node.is_element() || self.node.is_text_node()
+        self.node.is_text_node() || self.node.is_element()
     }
 }
 
+impl<'le> NodeInfo for ServoThreadSafeLayoutElement<'le> {
+    fn is_element(&self) -> bool {
+        self.pseudo == PseudoElementType::Normal
+    }
+
+    fn is_text_node(&self) -> bool {
```

Comment:
Probably we should `panic` here (or similar) given we're doing a totally unnecessary check.